convert2rhel/cert.py

```diff
@@ -20,56 +20,99 @@
-        except OSError as err:
-            loggerinst.critical("OSError({0}): {1}".format(err.errno, err.strerror))
+        if self.enabled:
+            return
+
+        target_path = os.path.join(self._target_cert_dir, self._cert_filename)
+        if os.path.exists(target_path):
+            loggerinst.info("Certificate already present at %s. Skipping copy." % target_path)
+            self.previously_installed = True
+        else:
+            try:
+                utils.mkdir_p(self._target_cert_dir)
+                shutil.copy2(self._source_cert_path, self._target_cert_dir)
+            except OSError as err:
+                loggerinst.critical("OSError({0}): {1}".format(err.errno, err.strerror))
 
-        loggerinst.info("Certificate %s copied to %s." % (self._cert_filename, self._target_cert_dir))
+            loggerinst.info("Certificate %s copied to %s." % (self._cert_filename, self._target_cert_dir))
 
-    def remove(self):
+        super(SystemCert, self).enable()
+
+    def restore(self):
         """Remove certificate (.pem), which was copied to system's cert dir."""
         loggerinst.task("Rollback: Remove installed RHSM certificate")
 
+        if self.enabled and not self.previously_installed:
+            self._restore()
+        else:
+            loggerinst.info("Certificate %s was present before conversion. Skipping removal." % self._cert_filename)
+
+        super(SystemCert, self).restore()
+
+    def _restore(self):
+        """The actual code to remove the certificate.  Done in a helper method so we can handle all
+        the cases where we do not want to remove and still run the base class's restore().
+        """
+        # Check whether any package owns the certificate file.  It is
+        # possible that a new rpm package was installed after we copied the
+        # certificate into place which owns the certificate.  If that is
+        # the case, then we don't want to remove it now.
+        output, code = utils.run_subprocess(["rpm", "-qf", self._target_cert_path])
+
+        # * If a file is owned by a package, rpm returns exit code 0 and prints
+        #   the package name to stdout.
+        # * If the file is not owned by a package, rpm returns exit code 1 and
+        #   prints a message about not being owned by any packagei to stdout.
+        # * If rpm encounters an error looking for the file (for instance, it
+        #   doesn't exist on the filesystem), then it returns exit code 1 and
+        #   prints the error message to stderr.
+        file_unowned = False
+        if code != 0:
+            if "not owned by any package" in output:
+                file_unowned = True
+            else:
+                loggerinst.warning(
+                    "Unable to determine if a package now owns" " %s. Skipping removal." % self._target_cert_path
+                )
+        else:
+            loggerinst.info(
+                "A package was installed which now owns the certificate"
```

Comment:
Unsure if this is better, just felt better to avoid now. Disregard if not
```suggestion
                "A package was installed that owns the certificate"
```